str2host.c

```diff
@@ -1800,3 +1800,282 @@ ldns_str2rdf_amtrelay(ldns_rdf **rd, const char *str)
+	ldns_buffer *str_buf = NULL;
+	char *key_str = NULL;
+	char *val_str = NULL;
+	char *item_str = NULL;
+	uint8_t *item_buffer = NULL;
+	ldns_rdf *item_rdf = NULL;
+
+	if(rd == NULL) {
+		return LDNS_STATUS_NULL;
+	}
+
+	str_buf = LDNS_MALLOC(ldns_buffer);
+	if(!str_buf) {
+		return LDNS_STATUS_MEM_ERR;
+	}
+	ldns_buffer_new_frm_data(str_buf, (char *)str, strlen(str));
+	if(ldns_buffer_status(str_buf) != LDNS_STATUS_OK) {
+		status = LDNS_STATUS_MEM_ERR;
+		goto error;
+	}
+
+	token = LDNS_XMALLOC(char, LDNS_MAX_RDFLEN);
+	if (!token) {
+		status = LDNS_STATUS_MEM_ERR;
+		goto error;
+	}
+
+	data = LDNS_XMALLOC(uint8_t, LDNS_MAX_RDFLEN);
+	if (!data) {
+		status = LDNS_STATUS_MEM_ERR;
+		goto error;
+	}
+	datap = data;
+
+	while (ldns_bget_token(str_buf, token, delimiters, LDNS_MAX_RDFLEN) > 0) {
+		//printf("AWS token %s\n", token);
+		key_str = LDNS_XMALLOC(char, strlen(token) + 1);
+		if (!key_str) {
+			status = LDNS_STATUS_MEM_ERR;
+			goto error;
+		}
+		val_str = LDNS_XMALLOC(char, strlen(token) + 1);
+		if (!val_str) {
+			status = LDNS_STATUS_MEM_ERR;
+			goto error;
+		}
+		char *equals_in_str = strchr(token, '=');
+		if (equals_in_str == NULL) {
+			// Just a key with no value, set value str to empty string
+			strcpy(key_str, token);
+			val_str[0] = '\0';
+		} else {
+			uint equals_idx = equals_in_str - token;
+			uint token_len = strlen(token);
+			if (equals_idx > 0 && equals_idx < strlen(token) - 1) {
+				// '=' is in the middle of the string, split key and value on it
+				memcpy(key_str, token, equals_idx);
+				key_str[equals_idx] = '\0';
+				// Strip off qoutes around the value if they are used
+				if (token[equals_idx+1] == '"' && token[token_len-1] == '"') {
+					if (token_len - equals_idx < 4) {
+						status = LDNS_STATUS_INVALID_SVCB_PARAM_VALUE;
+						goto error;
+					}
+					memcpy(val_str, equals_in_str + 2, token_len - equals_idx - 2);
+					val_str[token_len - equals_idx - 3] = '\0';
+				} else {
+					strcpy(val_str, equals_in_str + 1);
+				}
+			} else {
+				// '=' is at the beginning or end of the token, this is not allowed
+				status = LDNS_STATUS_INVALID_SVCB_PARAM_VALUE;
+				goto error;
+			}
+		}
+
+		int key = ldns_svcbparams_text2key(key_str);
+		if (key == -1) {
+			status = LDNS_STATUS_INVALID_SVCB_PARAM_KEY;
+			goto error;
+		}
+		ldns_write_uint16(datap, key);
+		datap += 2;
+		
+		// no-default-alpn must have no value, all other knonw types must have
+		// a non-zero length value, and unknown types may or may not have a value
+		if ((val_str[0] == '\0' && key != 2 && key <= 6) ||
+				(val_str[0] != '\0' && key == 2)) {
+			status = LDNS_STATUS_INVALID_SVCB_PARAM_VALUE;
+			goto error;
+		}
+		uint8_t *val_datap = datap;
+		datap += 2;
+
+		if (key == 0) {
+			// mandatory has comma separated list of key strings
+			char *start = val_str;
+			char *end = start + strlen(val_str);
+			bool escaped = false;
+			for (char *p = start; p <= end; p++) {
+				if (*p == '\\' && !escaped) {
```

Comment:
This was only meant to handle the case of an escaped comma.  But its not actually needed here since mandatory has the more restrictive set of allowed characters for key names.  So I will removed the part about escaped commas.